src/protocols/LoRaWAN/LoRaWAN.h

```diff
@@ -1016,94 +979,128 @@ class LoRaWANNode {
+    bool execMacCommand(uint8_t cid, uint8_t* optIn, uint8_t lenIn, uint8_t* optOut);
 
-    // parse a MAC command queue into a buffer (uplinks)
-    void macQueueToBuff(LoRaWANMacCommandQueue_t* queue, uint8_t* buffer);
+    // possible override for additional MAC commands that are not in the base specification
+    virtual bool derivedMacHandler(uint8_t cid, uint8_t* optIn, uint8_t lenIn, uint8_t* optOut);
 
-    // parse a MAC buffer into a command queue (downlinks)
-    void macBufftoQueue(LoRaWANMacCommandQueue_t* queue, uint8_t* buffer, uint8_t len);
+    void preprocessMacLinkAdr(uint8_t* mPtr, uint8_t cLen, uint8_t* mAdrOpt);
 
-    // push MAC command to queue, done by copy
-    int16_t pushMacCommand(LoRaWANMacCommand_t* cmd, LoRaWANMacCommandQueue_t* queue);
+    void postprocessMacLinkAdr(uint8_t* ack, uint8_t cLen);
 
-    // delete a specific MAC command from queue, indicated by the command ID
-    // if a payload pointer is supplied, this returns the payload of the MAC command
-    int16_t deleteMacCommand(uint8_t cid, LoRaWANMacCommandQueue_t* queue, uint8_t* payload = NULL);
+    int16_t getMacCommand(uint8_t cid, LoRaWANMacCommand_t* cmd);
 
-    // execute mac command, return the number of processed bytes for sequential processing
-    bool execMacCommand(LoRaWANMacCommand_t* cmd);
+    // possible override for additional MAC commands that are not in the base specification
+    virtual int16_t derivedMacFinder(uint8_t cid, LoRaWANMacCommand_t* cmd);
 
-    // apply a channel mask to a set of readily defined channels (dynamic bands only)
-    bool applyChannelMaskDyn(uint8_t chMaskCntl, uint16_t chMask);
+    //
+    int16_t getMacLen(uint8_t cid, uint8_t* len, uint8_t dir, bool inclusive = false);
 
-    // define or delete channels from a fixed set of channels (fixed bands only)
-    bool applyChannelMaskFix(uint8_t chMaskCntl, uint16_t chMask);
+    bool isPersistentMacCommand(uint8_t cid, uint8_t dir);
 
-    // get the payload length for a specific MAC command
-    uint8_t getMacPayloadLength(uint8_t cid);
-    
-    /*!
-      \brief Configures CSMA for LoRaWAN as per TR-13, LoRa Alliance.
-      \param backoffMax Num of BO slots to be decremented after DIFS phase. 0 to disable BO.
-      \param difsSlots Num of CADs to estimate a clear CH.
-      \param enableCSMA enable/disable CSMA for LoRaWAN.
-    */
-    void setCSMA(uint8_t backoffMax, uint8_t difsSlots, bool enableCSMA = false);
+    // push MAC command to queue, done by copy
+    int16_t pushMacCommand(uint8_t cid, uint8_t* cOcts, uint8_t* out, uint8_t* lenOut, uint8_t dir);
+
+    int16_t getMacPayload(uint8_t cid, uint8_t* in, uint8_t lenIn, uint8_t* out, uint8_t dir);
+
+    // delete a specific MAC command from queue, indicated by the command ID
+    int16_t deleteMacCommand(uint8_t cid, uint8_t* inOut, uint8_t* lenInOut, uint8_t dir);
+
+    void clearMacCommands(uint8_t* inOut, uint8_t* lenInOut, uint8_t dir);
```

Comment:
While these are protected members, it would still be good to have at least a short comment explaining the methods here, like the others have.